indicators/Adl/Adl.cs

```diff
@@ -7,11 +7,11 @@ namespace Skender.Stock.Indicators
     public static partial class Indicator
     {
         // ACCUMULATION / DISTRIBUTION LINE
```

Comment:
I'm not an expert on Interface patterns, so I might need some help understanding.  Why use `TQuote` instead of a simpler use of `IQuote`:

```
        // ACCUMULATION / DISTRIBUTION LINE
        public static IEnumerable<AdlResult> GetAdl(IEnumerable<IQuote> history, int? smaPeriod = null)
        {

            // clean quotes
            List<IQuote> historyList = history.Sort();

            // check parameters
            ValidateAdl(history, smaPeriod);

            // initialize
            List<AdlResult> results = new List<AdlResult>();
            decimal prevAdl = 0;

            // get money flow multiplier
            for (int i = 0; i < historyList.Count; i++)
            {
                IQuote h = historyList[i];
                [..]
```

Reply:
it's because covariance and contravariance
we are not passing a IQuote here but an IEnumeration<IQuote> and there are limits because of that
this post may contains useful information to you https://stackoverflow.com/questions/2033912/c-sharp-variance-problem-assigning-listderived-as-listbase
there are places that would require you to explicitly cast from a list of interfaces to a list of your concrete types (`ToList<YourClassThatImplementIQuote>`)

Using directly IQuote was my first attempt that I had in this branch, you can see the diffs
https://github.com/DaveSkender/Stock.Indicators/compare/master...MithrilMan:quote-to-iquote?expand=1

you can see that this way you are forced to work with IEnumerable<IQuote> while with generics you have typed list that uses the type of your IQuote implementation, so this solves the problem without requiring casts


About the documentation, I didn't have much time right now so I committed the most important pieces (and because of time constraint I didn't wanted to waste time on something that wasn't sure to be accepted/merged)

Reply:
Yah, it’s good to hold on docs.  I’m still trying to wrap my head around it.

Reply:
About the use of generics?
Generics are quite powerful and a core feature on c#, if you want an usecase about why the interface is useful, imagine having a feed of data coming from a 3rd party library (that have it's own model) then you are building a system where you have your own model because you can obtain data from different feeds for example, or because (it's always a good choice) you don't want to depend on 3rd party library models in case you want to swap them later, or provide alternative implementations for that part of software.
Now what you do is mapping the feed models into your own models (and if the 3rd party library doesn't allow you to specify your own model as a type you are forced to do the mapping.
Now it's your library turn, the user wants to use it to generate indicators on the data he has (in it's own models).
With current implementation he's forced to do another mapping into your Quote object, while with this modification, he has just to implement the IQuote interface on its own data model to map its own field into your required fields

Now he's able to feed, let say a `List<MyDataCandleStick>` into your indicator (where `MyDataCandleStick` is implementing IQuote) without having to spend time mapping from one type to another (less code, less computation, less memory allocation)

In a real time application, where data is feeded continuously (e.g. from a websocket that stream data), not having to perform all the mapping everytime is important (the alternative would be to keep a copy (so waste of memory) of the stream as a `IEnumerable<Quote>` or similar and just map the fresh data and append it to historical (like I'm actually doing) but it's far from efficient and force you to maintain more code too

For a simple usage of the library, nothing change for your users, they can use your Quote class as they are doing now, for advanced useage you allow programmers to reuse their model (just by implementing the interface)

Reply:
I get that part, a user being able to keep and use their model.  I'm mostly struggling with `TQuote` vs `IQuote` usage here and why the simpler `IQuote` won't work (or is somehow inferior).